### PR TITLE
302 m record trainee programme details

### PR DESCRIPTION
**Pull request by defong**
### Context
record trainee programme details

### Changes proposed in this pull request
Add record trainee programme details

### Guidance to review

image

image

image

### CODE REVIEW

**Comment by defong** [comment]
[[M] Record Trainee Programme Details](https://trello.com/c/sR6mTW3K/302-m-record-trainee-programme-details)

File: app/helpers/application_helper.rb [hunk 1]
```diff
@@ -1,2 +1,9 @@
 module ApplicationHelper
+  def boilerplate(array)
```
Comment (darokel) [inline app/helpers/application_helper.rb:2]: could we rename this to something a little more specific?

File: app/helpers/degrees_helper.rb [hunk 2]
```diff
@@ -1,4 +1,5 @@
 module DegreesHelper
+  include ApplicationHelper
   def hesa_degree_types_options
```
Comment (darokel) [inline app/helpers/degrees_helper.rb:3]: could we space these two lines out

File: app/models/programme_detail.rb [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+  attr_accessor :trainee, :subject, :main_age_range,
+                :additional_age_range, :day, :month, :year
+
+  delegate :id, :persisted?, to: :trainee
+
+  validates :subject, presence: true
+  validate :age_range_valid
+  validate :programme_start_date_valid
+
+  after_validation :update_trainee
```
Comment (darokel) [inline app/models/programme_detail.rb:15]: nice

File: spec/features/trainees/edit_programme_details_spec.rb [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    end
+  end
+
+  def then_i_see_error_messages
+    expect(page).to have_content(
+      I18n.t(
+        "activemodel.errors.models.programme_detail.attributes.subject.blank",
+      ),
+    )
+    expect(page).to have_content(I18n.t(
```
Comment (darokel) [inline spec/features/trainees/edit_programme_details_spec.rb:76]: could we align these two assertions the same way as above for consistency?

File: spec/models/programme_detail_spec.rb [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+  let(:trainee) { build(:trainee) }
+
+  subject { described_class.new(trainee: trainee) }
+
+  describe "validations" do
+    it { is_expected.to validate_presence_of(:subject) }
+
+    describe "custom" do
+      translation_key_prefix = "activemodel.errors.models.programme_detail.attributes"
+      before do
```
Comment (darokel) [inline spec/models/programme_detail_spec.rb:13]: could we add some spacing in this spec, some scenarios feel a bit squished

**Review by darokel — APPROVED** [review]
Code looks good overall 👍  I've left a few cosmetic comments (maybe just me being pedantic) which would be nice to do (pretty please 😄)

File: app/helpers/degrees_helper.rb (hunk 2 above)
Comment (defong) [reply app/helpers/degrees_helper.rb:4]: cool, made it less dense

File: spec/models/programme_detail_spec.rb (hunk 5 above)
Comment (defong) [reply spec/models/programme_detail_spec.rb:14]: cool, made it less dense

File: spec/features/trainees/edit_programme_details_spec.rb (hunk 4 above)
Comment (defong) [reply spec/features/trainees/edit_programme_details_spec.rb]: done

File: app/helpers/application_helper.rb (hunk 1 above)
Comment (defong) [reply app/helpers/application_helper.rb]: `to_options` next time we have one called `to_bar` :beer:

**Review by aliuk2012 — APPROVED** [review]

**Pull request merged**